tests/components/mysensors/test_gateway.py

```diff
@@ -0,0 +1,301 @@
+        with patch('socket.getaddrinfo'):
+            res = await async_setup_component(hass, 'mysensors', config)
+    assert res
+    await hass.async_block_till_done()
+    gateway = get_gateway(hass)
+    assert gateway.device == 'domain.com'
+    assert not gateway.sensors
+
+
+async def test_mqtt_device(hass, caplog, mock_mysensors, mqtt_mock):
+    """Test setup of mysensors with mqtt device."""
+    config = {'mysensors': {
+        'gateways': [{'device': 'mqtt'}], 'version': '2.0',
+        'persistence': False}}
+
+    res = await async_setup_component(hass, 'mysensors', config)
+    assert res
+    await hass.async_block_till_done()
+    gateway = get_gateway(hass)
+    assert gateway.device == 'mqtt'
+    assert not gateway.sensors
+
+    topic = 'test'
+    payload = 'test_payload'
+    qos = 0
+    retain = False
+
+    gateway.pub_callback(topic, payload, qos, retain)
+    await hass.async_block_till_done()
+
+    assert mqtt_mock.async_publish.call_count == 1
+    assert mqtt_mock.async_publish.call_args == call(
+        topic, payload, qos, retain)
+
+    calls = []
+
+    def sub_callback(*args):
+        """Mock the subscription callback."""
+        calls.append(args)
+
+    gateway.sub_callback(topic, sub_callback, qos)
+    await hass.async_block_till_done()
+    async_fire_mqtt_message(hass, topic, payload, qos, retain)
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()  # not sure why we need two?
+
+    assert len(calls) == 1
+    assert calls[0] == (topic, payload, qos)
+
+
+async def test_persistence(
+        hass, mock_isdevice, mock_discover_platform, mock_mysensors,
+        mock_gw_ready):
+    """Test MySensors gateway persistence."""
+    config = {'mysensors': {
+        'gateways': [{'device': DEVICE, 'persistence_file': 'test.json'}],
+        'version': '2.0', 'persistence': True}}
+    res = await async_setup_component(hass, 'mysensors', config)
+    assert res
+    gateway = get_gateway(hass)
+    gateway.sensors[1] = MockNode(1)
+    child_type = gateway.const.Presentation.S_TEMP
+    value_type = gateway.const.SetReq.V_TEMP
+    gateway.sensors[1].children[1] = child = MockChild(1, child_type)
+    payload = '20.0'
+    child.values[value_type] = payload
+    new_dev_ids = [(id(gateway), 1, child.id, value_type)]
+    await hass.async_block_till_done()
+    assert mock_discover_platform.call_count == 1
+    assert mock_discover_platform.call_args == call(
+        hass, 'sensor', mysensors_comp.DOMAIN,
+        {'devices': new_dev_ids, 'name': mysensors_comp.DOMAIN})
+
+
+async def test_gateway_not_ready(hass, caplog, mock_isdevice, mock_mysensors):
+    """Test gateway not ready."""
+    called = False
+
+    async def fut():
+        """Fake future that raises asyncio.TimeoutError."""
+        nonlocal called
+        called = True
+        raise asyncio.TimeoutError
+
+    with patch('homeassistant.components.mysensors.gateway.asyncio.Future',
```

Comment:
Alternative approach is to use `side_effect=asyncio.TimeoutError` instead of `return_value`, and the exception will be raised. Now save the output of the `with` statement with `as mock_future` and you can now assert that `mock_future` was called.

btw, mocking out something so deep in the core of Python seems a bit aggressive. Could you mock something around it ?

Reply:
I don't think a standard mock can be awaited.

Reply:
It can't, but if the creation of the Future already raises a TimeoutError, shouldn't that have the same effect for testing?

Reply:
It's not the creation of the future that should raise, it's awaiting the future that should raise.

Reply:
Ah I see. Not that it helps a lot, but you _could_ always use [`setException`](https://docs.python.org/3/library/asyncio-task.html#asyncio.Future.set_exception) on a future, however this will work too. It's just a bit error prone as any Future in the system will be impacted by this patch.

Reply:
Ok, I'll look at that.